### ISSUE TITLE
[Editor2] New map template

### ISSUE THREAD

**Issue by LordSk** [teeworlds/teeworlds#2577]
On [File > New] open a popup to select a map template to start off of.
A template consists of:

* A theme (grass, jungle, desert, snow)
* A size (small, medium, large)

Once the template is selected, a new map is created based on it.

* The sky quad should have the theme colour
* All theme images should be loaded (grass_main, grass_doodads, clouds)
* A tilemap layer of the theme's size should be created (and have the border filled in?)
* The game layer should have a few (maybe only one) player spawn(s)

Now the easiest way I can think of doing all these is to create a few map files (such as grass_medium.map), and load them in. It would also allow us to do more fancy stuff, like adding a sun and clouds by default.
We could also auto detect map templates by scanning these map files. So egypt_X.map files would add the egypt theme.

It's also possible to do a hybrid approach by generating the game layer and tilemap layer ourselves (as it is easy to do) and only rely on the map files for the background and image loading.
Thoughts?

**Comment by Sonix-** [comment]
I had a similar idea some time ago about providing basic map themes/assets like backgrounds so an user could simply load them in. [teeworlds-maps issue](https://github.com/teeworlds/teeworlds-maps/issues/20)


> On [File > New] open a popup to select a map template to start off of.
> A template consists of:
> 
> * A theme (grass, jungle, desert, snow)

blank/none and custom should also be an option

> * A size (small, medium, large)
> 
> Once the template is selected, a new map is created based on it.
> 
> * The sky quad should have the theme colour
> * All theme images should be loaded (grass_main, grass_doodads, clouds)
> * A tilemap layer of the theme's size should be created (and have the border filled in?)

I wouldn't fill the border automatically

> * The game layer should have a few (maybe only one) player spawn(s)
> 
> Now the easiest way I can think of doing all these is to create a few map files (such as grass_medium.map), and load them in. It would also allow us to do more fancy stuff, like adding a sun and clouds by default.

We could create a folder in /maps called presets/templates and provide basic and generic map files there containing just the background.
Maps shouldn't be bound by size (_medium _small) instead just increase automatically the layer to the size the user has chosen.

**Comment by LordSk** [comment]
> I had a similar idea some time ago about providing basic map themes/assets like backgrounds so an user could simply load them in. teeworlds-maps issue

I was not aware of that, looks like our objectives align 😃.

> blank/none and custom should also be an option

Yep.

> I wouldn't fill the border automatically

Some kind of starting point would be nice though, right? Having a tilemap layer ready that you can just add to. Even if it is a single platform with a spawn on top of it.

> We could create a folder in /maps called presets/templates and provide basic and generic map files there containing just the background.

Yes.

> Maps shouldn't be bound by size (_medium _small) instead just increase automatically the layer to the size the user has chosen.

So I have a plan to make the whole layer size stuff a thing of the past. You might be referring to the same idea, but just to be clear, the small/medium/large setting is just a starting point and does not lock the size of the map (even with the current system).

To summarise: hybrid approach +1

**Comment by Sonix-** [comment]
store templates in maps/templates? grass, desert, winter, jungle

Usage could be:  
click New map --> pop up dialog "Choose a template" or use default --> If click choose template --> Desert, Grass, Jungle, Winter  --> once chosen select size dialog "small, medium, large" --> if small do 50x50 layers, if medium 100x100 if large 150x100 or 150x150.. 

Instead if u click on default it uses the same preset as the current editor.

**Comment by Learath2** [comment]
From what I gather it's trivial to arbitrarily size these maps, that's why I used an input for the size. The templates should include a quad layer to begin with, I'm not sure if we want to do anything with other layers, maybe set up a tile layer with the correct tileset but do we add a platform or something?

If we want a platform we could generate a platform programatically and have automapper tile it, or we could have a platform in the template map itself.

**Comment by Sonix-** [comment]
I wouldnt make an user bother with finding the right size from the start, he can use one of the 3 presets and then adjust to his desires

**Comment by LordSk** [comment]
Alright let's do this then:

* Make a new popup
	- Display a template preview image (data/editor/templates/$NAME_preview.png).
	- Selecting 'Empty' hides the preview and loads with the default empty map ( CEditorMap2::LoadDefault() )

![image](https://user-images.githubusercontent.com/294603/80861834-5b596600-8c71-11ea-9420-23a5c33f54ab.png)
Colours are not accurate here, use the editor colour 'Style' variables.
*Just a reminder: UI is temporary, I'm just going for something simple here.*

* Fetch templates
	- List files in 'data/editor/templates' once on editor init.
	- Create a template-preview pair for each $NAME_template.map
	- Load the preview image ($NAME_preview.png) if found

* Make a 'create map from template' function which takes a name and size as argument (so we can make a command to call it, like: `new grass medium`)
	- Load a template map (data/editor/templates/$NAME_template.map)
	- Clear the map path
	- Resize the Game Layer and main Tilemap layer to the chosen size
	- Create a simple rectangle platform and automap it (if applicable)
	- Make the platform solid (game layer collision) and add a neutral spawn on top of it


# Template maps content
* Quad background
* Sky stuff, snow or anything in the background is good as well
* All the template images added (main, doodads, etc)
* A tilemap on top of the game layer with the main image applied to it (example: grass_main)

# Template preview
* Size: to be determined
